### PR TITLE
Creation of menus on the edit navigation screen

### LINKED ISSUE #21281: Allow creation/deletion of Navigation Menus on the Navigation screen.
**Is your feature request related to a problem? Please describe.**
#21036 implemented the experimental nav screen. Currently a user can select an existing menu created through the existing menu screen, but cannot create or delete menus.

This should be possible from the screen.

**Describe the solution you'd like**
- [ ] A button for creating new menus.
- [x] A button for deleting menus (with a confirmation stage).

### PR DESCRIPTION
**Pull request by talldan**
## Description
Closes #21281

Allows creation of new menus from the experimental navigation menu screen.

This is fairly rough and ready, but I think good enough for a first draft.

### First menu
1. Visit the page when there are no menus
2. Create a new menu

### With existing menus
1. Visit the page when there are menus
2. Click the Create Menu button
3. Create a new menu

## Screenshots 
Screenshot 2020-05-14 at 6 11 24 pm

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by draganescu** [comment]
👋 I tested this locally and noted the following:

- I should have a cancel link to get out of the create menu form and return to where I was
- there is a flicker when creating menus, probably everything gets re-rendered
- if I create a menu with a name that already exists, the menu does not get created but I'm not notified
- we should rebase this to include the menu locations to see if the new menu appears there as well (currently for me master crashes)

**Comment by talldan** [comment]
Thanks for testing @draganescu.

I'll work on validation, cancelling menu creation, and also have a fix for that error (that I'll move into a separate PR as well).

The flickering is a tricky one to solve and I might have to defer to a separate PR. It seems to be related to `core/data` refetching menus again after a menu creation.

It might be possible to use some smoke and mirrors to hide this, maybe by adding a debounce to the spinner.

File: packages/edit-navigation/src/components/menus-editor/create-menu-panel.js
Comment (talldan) [inline packages/edit-navigation/src/components/menus-editor/create-menu-panel.js:46]: This doesn't seem to work. Maybe the component is unmounted before the effect gets a chance to fire.

File: packages/edit-navigation/src/components/menus-editor/create-menu-panel.js
Comment (talldan) [inline packages/edit-navigation/src/components/menus-editor/create-menu-panel.js:118]: I'd hoped this would return focus to the 'Create a new menu' button, but I think because that button isn't present in the DOM and focusable, it fails. Not sure what a good way to solve that would be.

File: packages/edit-navigation/src/components/menus-editor/index.js
Comment (talldan) [inline packages/edit-navigation/src/components/menus-editor/index.js:74]: Not a huge fan of setting the menu id to `undefined` to indicate a menu is being created. Open to suggestions for improving this. Potentially could use some state, e.g.
```
editMode: 'CREATE_MENU'
```

File: packages/edit-navigation/src/components/menus-editor/index.js
Comment (talldan) [inline packages/edit-navigation/src/components/menus-editor/index.js:105]: This fixes a separate bug that when you delete the last menu, the menuId still refers to it.

**Comment by talldan** [comment]
Not too far off with this, though there are a few issues I've outlined in my comments above. Would welcome further testing, reviews, and opinions!

**Comment by draganescu** [comment]
One other functional tid bit @talldan, it would be nice that after menu creation the selected one is the newly created menu.

File: packages/edit-navigation/src/components/menu-editor/block-editor-panel.js
Comment (adamziel) [inline packages/edit-navigation/src/components/menu-editor/block-editor-panel.js:55]: Should this component just return null when `hasMenu === false`?

File: packages/edit-navigation/src/components/menu-editor/navigation-structure-panel.js
Comment (adamziel) [inline packages/edit-navigation/src/components/menu-editor/navigation-structure-panel.js:27]: Same as above - should this component just return null when `showNavigationStructure === false`?

File: packages/edit-navigation/src/components/menu-editor/style.scss
Comment (adamziel) [inline packages/edit-navigation/src/components/menu-editor/style.scss:72]: It seems like we're using either 6px or 8px for this kind of margin in other places.

File: packages/edit-navigation/src/components/menus-editor/create-menu-panel.js
Comment (adamziel) [inline packages/edit-navigation/src/components/menus-editor/create-menu-panel.js:74]: Something like this should take care of the notifications.
```suggestion
			try {
				await saveMenu( { name: menuName } );
				createInfoNotice( __( 'Menu created' ), {
					type: 'snackbar',
					isDismissible: true,
				} );
			} catch (err) {
				createErrorNotice( __( 'Menu creation failed' ), {
					type: 'snackbar',
					isDismissible: true,
				} );
			}
```

**Comment by adamziel** [comment]
Aside of the things already mentioned in comments, it seems to work pretty well. Good job @talldan !

**Comment by paaljoachim** [comment]
I am testing the PR at gutenberg.run.

This is my initial test of the Navigation (beta) screen, and I am very impressed with all the work that has been done up to this point. Very nice!

The process of creating a menu.
1. Click **Create a new menu**
Screen Shot 2020-05-14 at 16 35 11

2.
A new screen opens.
Screen Shot 2020-05-14 at 15 59 30

It surprised me that a new screen was seen. 


To me it would seem that the accordion would be a part of the general **Edit Navigation** screen. 

Nav-Create-new-menu-1

Open accordion.
Nav-screen-accordion-open

Another idea. Click "Create a new menu" and the accordion is seen below along with the other options.
Nav-same-screen

Here the user clicks "Create a new menu" and it opens in the same screen. It does not open in another "Edit Navigation" screen.

User fills out the name "Top nav" and clicks "Create menu" and the "Edit Navigation" screen is auto updated.
Screen Shot 2020-05-14 at 16 34 01

I am not saying the above is a good idea, but I wanted to bring up the initial brain storming that hit me.

File: packages/edit-navigation/src/components/menu-editor/block-editor-panel.js
Comment (talldan) [reply packages/edit-navigation/src/components/menu-editor/block-editor-panel.js]: I ended up removing that and going with the suggestion of @paaljoachim below and showing the create panel alongside other panels.

File: packages/edit-navigation/src/components/menu-editor/navigation-structure-panel.js
Comment (talldan) [reply packages/edit-navigation/src/components/menu-editor/navigation-structure-panel.js:23]: I changed this back to how it is now for the same reason as above, the panel will show at the same time as the create panel.

File: packages/edit-navigation/src/components/menus-editor/create-menu-panel.js
Comment (talldan) [reply packages/edit-navigation/src/components/menus-editor/create-menu-panel.js]: Thanks, that works nicely 👍

**Comment by talldan** [comment]
I think I've fixed most of the issues now. I agree with @paaljoachim's suggestion of having the panel appear alongside the other UI, so it works similarly that now. That also inadvertently fixed the focus management issues 🎉 

Also managed to fix the glitchy page updates that were occurring when creating a menu and the success/error notifications. It feels a lot smoother now.

File: packages/edit-navigation/src/components/menus-editor/create-menu-panel.js
Comment (adamziel) [inline packages/edit-navigation/src/components/menus-editor/create-menu-panel.js:56]:
```suggestion
				createErrorNotice( message, { id: noticeId } );
				setIsCreatingMenu( false );
```

File: packages/edit-navigation/src/components/menus-editor/create-menu-panel.js
Comment (adamziel) [inline packages/edit-navigation/src/components/menus-editor/create-menu-panel.js:37]:
```suggestion
			
```

File: packages/edit-navigation/src/components/menus-editor/create-menu-panel.js
Comment (adamziel) [inline packages/edit-navigation/src/components/menus-editor/create-menu-panel.js:59]:
```suggestion
			try {
				setIsCreatingMenu( true );
```

File: packages/edit-navigation/src/components/menus-editor/create-menu-panel.js
Comment (adamziel) [inline packages/edit-navigation/src/components/menus-editor/create-menu-panel.js:65]:
```suggestion
```

File: packages/edit-navigation/src/components/menus-editor/create-menu-panel.js
Comment (adamziel) [inline packages/edit-navigation/src/components/menus-editor/create-menu-panel.js:74]:
```suggestion
```

File: packages/edit-navigation/src/components/menus-editor/create-menu-panel.js
Comment (adamziel) [inline packages/edit-navigation/src/components/menus-editor/create-menu-panel.js:75]:
```suggestion
			} finally {
				setIsCreatingMenu( false );
			}
```

File: packages/edit-navigation/src/components/menus-editor/index.js
Comment (adamziel) [inline packages/edit-navigation/src/components/menus-editor/index.js:38]: Nit: does it need to depend on `hasCompletedFirstLoad`?

**Review by adamziel — APPROVED** [review]
@talldan the glitchy behavior is gone, good job on this one! I left a few minor comments but nothing overly big.

The ones referring to the progress state will fix this problem where you get stuck in a progress state in some cases (but the submit button continues to work):

Zrzut ekranu 2020-05-15 o 12 28 22

Also this is not specific to this PR, but I totally missed the notification, I think we should make them more visible cc @karmatosed @mapk:

Zrzut ekranu 2020-05-15 o 12 28 46

Another design thing is that I think we can do better in the future with how the empty state feels right after a new menu is created:

Zrzut ekranu 2020-05-15 o 12 30 54

**Comment by paaljoachim** [comment]
I went ahead and tested at gutenberg.run again.
Experimental settings: Turned on Enable Navigation screen.
Then went to the screen to create a menu. I tried making a menu and clicked Create Menu but nothing happened. So I again tried to create a menu. I then saw this:
Screen Shot 2020-05-15 at 12 37 36


Inspect:
{"code":"rest_cannot_view","message":"Sorry, you cannot view these menus, unless you have access to permission edit them. ","data":{"status":401}}

Screen Shot 2020-05-15 at 12 39 08

**Comment by paaljoachim** [comment]
Hey @adamziel 

Notifications in general show up on the bottom left. I see notifications as an extra visual mark in addition to an action that has just been taken by the user. In this case an extra visual mark that one has created a menu in addition to create menu action that is seen. (Which I do not have access to yet because of the above mentioned error. Or else I would add some screenshots and/or animated gif. I might update this post later on.)

**Comment by talldan** [comment]
@paaljoachim Thanks for testing. Guessing the error is related to permissions when using gutenberg.run. But the error messages should be better. 

It looks like the try/catch was only catching JavaScript errors, and for REST/entity errors the `getLastEntitySaveError` selector should be used, so I've implemented that, and the correct error message is now shown.

I've applied all the other suggestions and will merge this when tests pass.

I also discovered #22404 while working on this, that can cause the created menu not to be selected if there are more than 10 menus.

**Pull request merged**
